### PR TITLE
Start working on CLI2 update for NEU

### PR DESCRIPTION
**Pull request by tekknolagi**

### CODE REVIEW

File: assignments/02-cli-constructive.md [hunk 1]
```diff
@@ -3,77 +3,82 @@
...
+These languages all have one thing in common: they are predominantly compiled
+before being executed[^interpreters]. This two-step tango means that every time
+you modify your program, you have to compile it before running it.
+
+[^interpreters]: We say *predominantly* because the evaluation strategy is not
+    necessarily part of the language! The language is the abstract concept and
+    the implementation makes it go zoom. There are C and Rust interpreters and
+    there are (for example) Ruby and Python compilers.
+
+That's not fun, especially if it requires some kind of arcane build flags that
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:18]:
```suggestion
That's not fun, especially if it requires arcane compiler flags that
```

File: assignments/02-cli-constructive.md [hunk 2]
```diff
@@ -3,77 +3,82 @@
...
+are hard to remember. Or maybe it has a lot of files and your build command is
+getting unwieldy. Either way, you are going to solve your own problems today by
+*writing a build system*.
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:21]:
```suggestion
*writing a build script*.
```

less scary?

File: assignments/02-cli-constructive.md [hunk 3]
```diff
@@ -3,77 +3,82 @@
...
+
+You need not have written C code ever before starting this part of the
+assignment. We will provide sample code to you so that you can test your build
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:24]:
```suggestion
This part of the assignment uses the C compiler as an example, but it doesn't
require you to write C: we provide sample C code you can use to test your build
```

File: assignments/02-cli-constructive.md [hunk 4]
```diff
@@ -3,77 +3,82 @@
...
+* We can update individual C files without recompiling other C files
+
+...
+
+### Requirements
+
+> Note: The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+> "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
+> document are to be interpreted as described in [RFC
+> 2119](https://datatracker.ietf.org/doc/html/rfc2119).
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:74]: Needed? I'm pretty sure students know what "must" means when it occurs in an assignment, and that's the only one you use.

File: assignments/02-cli-constructive.md [hunk 5]
```diff
@@ -3,77 +3,82 @@
...
-daily.2021-09-21_0010             every_four_hours.2021-09-23_2005
-daily.2021-09-22_0010             every_four_hours.2021-09-24_0005
-daily.2021-09-23_0010             weekly.2021-09-05_0015
-daily.2021-09-24_0010             weekly.2021-09-12_0015
-every_four_hours.2021-09-23_0405  weekly.2021-09-19_0015
-every_four_hours.2021-09-23_0805
-$ 
+
+If only it were that simple for your assignment. We'll get more into all this
+later in the term, but this program is not ideal: it builds `foo.c` into `foo`
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:38]: I might try to reword this, but I'd need to think about it more. Will come back to this comment later.

File: assignments/02-cli-constructive.md [hunk 6]
```diff
@@ -6,40 +6,41 @@
...
 
-You need not have written C code ever before starting this part of the
-assignment. We will provide sample code to you so that you can test your build
+This part of the assignment uses the C compiler as an example, but it doesn't
+require you to write C: we provide sample C code you can use to test your build
 system. Later, though, you *will* write some C code and integrate it into your
 build system.
 
-Your job is to write a program in Bash, `build.sh`, that compiles some C
+Your job is to write a program in Bash or Sh, `build.sh`, that compiles some C
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:29]: I don't think we use the term "Sh" in notes or elsewhere.

File: assignments/02-cli-constructive.md [hunk 7]
```diff
@@ -6,40 +6,41 @@
...
 
 If only it were that simple for your assignment. We'll get more into all this
-later in the term, but this program is not ideal: it builds `foo.c` into `foo`
-every time `build.sh` is executed, even if `foo.c` has not changed. Your
-program must instead only rebuild `foo` if any of its dependencies have
-changed. In addition, your program must build intermediate object files for
-each C file.
+later in the term, but this program is not ideal: it builds `foo.c` and `rng.c`
+into `foo` every time `build.sh` is executed, even if `foo.c` and `rng.c` have
+not changed. Your program must instead only rebuild `foo` if any of its
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:41]:
```suggestion
not changed. The script you submit, in contrast, must only rebuild `foo` if any of its
```

File: assignments/02-cli-constructive.md [hunk 8]
```diff
@@ -55,63 +56,101 @@ $ touch rng.c
...
-...
+* We can update individual C files without recompiling other C files (we do
+  need to link them at the end, though)
+* By "changed", we mean that the file's modification time is newer than its
+  target; the file contents may not have changed
+
+This is a pretty standard set of features for build systems. Many people have
+put many years into making build systems great. Because we are dealing with
+small code and don't have a lot of strange requirements, you are only going to
+peek a little bit into that yawning abyss. We hope that after writing your
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:76]: I'd cut "We hope …" and everything after. I think referencing Make will be lost on them at this point in the course.

File: assignments/02-cli-constructive.md [hunk 9]
```diff
@@ -55,63 +56,101 @@ $ touch rng.c
...
+
 > Note: The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
 > "SHOULD", "SHOULD NOT", "RECOMMENDED",  "MAY", and "OPTIONAL" in this
 > document are to be interpreted as described in [RFC
 > 2119](https://datatracker.ietf.org/doc/html/rfc2119).
 
 You MUST:
 
+* Be able to build `foo` by running `./build.sh`
+* Only recompile a target if any of its dependencies' modified-time (mtime) is
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:98]: I think this requirement (rebuilding based on modification time) warrants a paragraph in the text above. We are not yet at BLD and so the fact that build systems work like this won't be evident to many students.

Comment (tchebb) [reply assignments/02-cli-constructive.md:98]: Also maybe rephrase as "if at least one of its dependencies has been modified more recently than it"

File: assignments/02-cli-constructive.md [hunk 10]
```diff
@@ -55,63 +56,101 @@ $ touch rng.c
...
-temporary files that don't need to stick around. We encourage you to remove any
-temporary files you use once your script finishes, but you will not lose points
-if you don't.
-
-## Using syscalls: write your own ls!
+* Use the error-handling practices we encouraged in lecture
+
+You MUST NOT:
+
+* Shell out to `make` or `watchman` or `entr` or any other software that would
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:108]: s/Shell out/Invoke/? A little less jargon-y.

File: assignments/02-cli-constructive.md [hunk 11]
```diff
@@ -55,63 +56,101 @@ $ touch rng.c
...
+* Define functions to make your code more readable
+* Support modifying the build by setting `CC`, `CFLAGS`, `LDFLAGS`, and other
+  assorted environment variables
+* Write data to temporary files as part of your build script. If you do, you
+  MUST:
+  * Use the `mktemp` utility to create them
+  * Automatically clean up the temporary files when the build script finishes
+
+In general, you MAY go above and beyond if you want to, as long as your
+extended program still meets the requirements.
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:148]: Should clarify if this will result in extra credit.

File: assignments/02-cli-constructive.md (hunk 8 above)
Comment (tekknolagi) [reply assignments/02-cli-constructive.md:98]: :(

File: assignments/02-cli-constructive.md [hunk 12]
```diff
@@ -55,63 +56,101 @@ $ touch rng.c
...
+* Shell out to `make` or `watchman` or `entr` or any other software that would
+  do substantive work for you
+* Modify any of the C source files we provide to you
+
+You MAY:
+
+* Print the commands executed by your script as they are executed
+* Take an optional argument to `build.sh` to specify what target to build
+* Use (small) standard Unix utilities. When in doubt, ask on Piazza
+* Define functions to make your code more readable
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:139]: Should this one be a SHOULD?

File: assignments/02-cli-constructive.md (hunk 10 above)
Comment (tekknolagi) [reply assignments/02-cli-constructive.md:130]: use

File: assignments/02-cli-constructive.md [hunk 13]
```diff
@@ -55,63 +56,101 @@ $ touch rng.c
...
+* Support modifying the build by setting `CC`, `CFLAGS`, `LDFLAGS`, and other
+  assorted environment variables
+* Write data to temporary files as part of your build script. If you do, you
+  MUST:
+  * Use the `mktemp` utility to create them
+  * Automatically clean up the temporary files when the build script finishes
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:148]: How strictly will you grade this in the rubric? If some command in their script fails unexpectedly and `set -e` bails, do the files still need to be deleted? What about if you Ctrl-C the script?

Basically asking if they need to know how to use `trap` to meet this requirement or not.

File: assignments/02-cli-constructive.md [hunk 14]
```diff
@@ -120,11 +159,24 @@ which is a set of function implementations that are available to any C program
...
+    "file" is often used to refer specifically to a *regular file*, which is
+    the "normal" kind of file that holds data and shows up in `ls` with a type
+    of `-`. However, the strict POSIX definition of "file" encompasses not only
+    regular files but also directories, symlinks, devices, and every other
+    thing that can go inside a directory. For this assignment, we are referring
+    to the POSIX definition whenever we say "file." Your implementation of
+    `myls.c` should consider directories, symlinks, and all other types of file
+    when producing their output.
+
+Now, onto implementation notes. We suggest looking into a couple of different
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:201]:
```suggestion
Now, on to implementation notes. We suggest looking into a couple of different
```

File: assignments/02-cli-constructive.md [hunk 15]
```diff
@@ -174,14 +226,18 @@ As an extension, you may allow the user to omit the argument and have your
...
 required for full marks.
 
+This is a small program. You may feel tempted to copy it off the internet or
+classmate. **Don't.** You are robbing yourself of learning. We have given you
+enough function names to get started. Check out their manual pages.
+
 ## Using `build.sh` to compile `myls.c`
 
-...
+Now integrate by adding `myls` as a target to `build.sh`! You MUST be able to
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:258]: Should this be an official "Part 3"? Seems easy to miss otherwise.

File: assignments/02-cli-constructive.md [hunk 16]
```diff
@@ -3,6 +3,28 @@
 
 # Homework 2: CLI, Constructive
 
+This is the first of four "constructive" assignments---assigments where we ask
+you to write a program. The point of such assignments is to get you more
+comfortable building things in a safe environment so that later, in a more
+"real" setting, you believe in yourself. While we are somewhat biased, we also
+think they can even be thought of as fun.
```
Comment (tchebb) [inline assignments/02-cli-constructive.md:10]:
```suggestion
This is the first of four "constructive" assignments---most of which ask
you to write a small amount of code. While writing that code, you will likely
make mistakes that highlight gaps in your understanding and require you to
fill those gaps by referencing lecture notes or documentation. Not only is that
expected, it's the main reason we created these assignments. The programs
you produce are secondary to the experience you gain while producing them.

In the past couple of years, it has gotten dramatically easier to bypass the
trial-and-error portion of a programming assignment by generating a program
using a large language model like ChatGPT or Copilot. Although such tools
have their place, they do not in general help you build understanding, which is
the primary purpose of these assignments.
```

File: assignments/02-cli-constructive.md (hunk 9 above)
Comment (tekknolagi) [reply assignments/02-cli-constructive.md:130]: Added note about mtime
